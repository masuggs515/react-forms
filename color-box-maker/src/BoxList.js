import React, { useState } from 'react'
import Box from './Box'
import NewBoxForm from './NewBoxForm'
import {v4 as uuid} from 'uuid'
import './BoxList.css'

const BoxList = () => {
    const INITIAL_LIST = [];

    const [boxes, setBoxes] = useState(INITIAL_LIST);

    const removeBox = box => {
        setBoxes(boxes.filter(b=> b.id !== box));
    };

    const addBox = (boxData) =>{
        setBoxes(boxes => [...boxes, {...boxData, id: uuid()}])
    }

    const finalBoxes = boxes.map(({id, color, height, width}) => <Box
    id={id}
    color={color}
    height={+height}
    width={+width} 
    removeBox={removeBox}
    addBox={addBox}
    key={id}/>
)


    return (
        <div className='BoxList'>
            
            <h1>Add Boxes</h1>
            <NewBoxForm addBox={addBox} />
                {finalBoxes}
        
        </ div>
    )
}

export default BoxList;