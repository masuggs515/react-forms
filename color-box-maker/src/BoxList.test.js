import { fireEvent, getByLabelText, getByText, render } from '@testing-library/react';
import BoxList from './BoxList';

const addBox = (box, height='300', width='400', color='blue') =>{
    const heightInput = box.getByLabelText("Box Height:");
    const widthInput = box.getByLabelText("Box Width:");
    const backgroundInput = box.getByLabelText("Box Color:");
    fireEvent.change(backgroundInput, { target: { value: color } });
    fireEvent.change(widthInput, { target: { value: width } });
    fireEvent.change(heightInput, { target: { value: height } });
    const btn = box.getByText('Add Box');
    fireEvent.click(btn);
};


it('renders', () => {
  render(<BoxList />);
});

it('matches snapshot', ()=>{
    const {asFragment} = render(<BoxList />);
    expect(asFragment()).toMatchSnapshot();
});

it('should add box', ()=>{
    const box = render(<BoxList />);

    expect(box.queryByText('X')).not.toBeInTheDocument();

    addBox(box);
    const delBtn = box.getByText('X');
    expect(delBtn).toBeInTheDocument();
        
    expect(delBtn.parentElement).toHaveStyle(`
    width: 400px;
    height: 300px;
    background-color: blue;
  `);
});

it('should remove box', ()=>{
    const box = render(<BoxList />);
    addBox(box);

    const delBtn = box.getByText('X');
    fireEvent.click(delBtn);
    expect(delBtn).not.toBeInTheDocument();
});