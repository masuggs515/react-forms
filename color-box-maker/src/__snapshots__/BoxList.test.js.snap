// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="BoxList"
  >
    <h1>
      Add Boxes
    </h1>
    <form
      class="NewBoxForm"
    >
      <label
        for="color"
      >
        Box Color:
      </label>
      <input
        id="color"
        name="color"
        placeholder="Enter Color..."
        type="text"
        value=""
      />
      <label
        for="height"
      >
        Box Height:
      </label>
      <input
        id="height"
        name="height"
        placeholder="Enter Height..."
        type="number"
        value=""
      />
      <label
        for="width"
      >
        Box Width:
      </label>
      <input
        id="width"
        name="width"
        placeholder="Enter Width..."
        type="number"
        value=""
      />
      <button>
        Add Box
      </button>
    </form>
  </div>
</DocumentFragment>
`;
